File: app/src/main/java/de/xikolo/controllers/channels/ChannelDetailsActivity.java
package de.xikolo.controllers.channels;

import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.app.NavUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewStub;
import android.widget.ImageView;

import com.yatatsu.autobundle.AutoBundleField;

import butterknife.BindView;
import de.xikolo.R;
import de.xikolo.config.GlideApp;
import de.xikolo.controllers.base.BaseActivity;
import de.xikolo.controllers.helper.CollapsingToolbarHelper;
import de.xikolo.models.Channel;
import de.xikolo.utils.ShareUtil;

public class ChannelDetailsActivity extends BaseActivity {

    public static final String TAG = ChannelDetailsActivity.class.getSimpleName();

    @AutoBundleField String channelId;

    @AutoBundleField(required = false) boolean scrollToCourses = false;

    @BindView(R.id.toolbar_image) ImageView imageView;
    @BindView(R.id.appbar) AppBarLayout appBarLayout;
    @BindView(R.id.collapsing_toolbar) CollapsingToolbarLayout collapsingToolbar;
    @BindView(R.id.stub_bottom) ViewStub stubBottom;
    @BindView(R.id.scrim_top) View scrimTop;
    @BindView(R.id.scrim_bottom) View scrimBottom;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_blank_collapsing);
        setupActionBar(true);
        enableOfflineModeToolbar(false);

        Channel channel = Channel.get(channelId);

        setTitle(channel.title);

        int color = channel.getColorOrDefault();
        collapsingToolbar.setContentScrimColor(color);
        collapsingToolbar.setBackgroundColor(color);
        collapsingToolbar.setStatusBarScrimColor(color);

        String tag = "content";

        if (channel.imageUrl != null) {
            GlideApp.with(this).load(channel.imageUrl).into(imageView);
        } else {
            CollapsingToolbarHelper.lockCollapsingToolbar(channel.title, appBarLayout, collapsingToolbar, toolbar, scrimTop, scrimBottom);
        }

        FragmentManager fragmentManager = getSupportFragmentManager();
        if (fragmentManager.findFragmentByTag(tag) == null) {
            final ChannelDetailsFragment fragment = ChannelDetailsFragmentAutoBundle.builder(channelId).scrollToCourses(scrollToCourses).build();
            FragmentTransaction transaction = fragmentManager.beginTransaction();
            transaction.replace(R.id.content, fragment, tag);
            transaction.commit();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        switch (itemId) {
            case android.R.id.home:
                NavUtils.navigateUpFromSameTask(this);
                return true;
            case R.id.action_share:
                ShareUtil.shareCourseLink(this, channelId);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/de/xikolo/controllers/channels/ChannelDetailsActivity.kt
package de.xikolo.controllers.channels

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.core.app.NavUtils
import com.yatatsu.autobundle.AutoBundleField
import de.xikolo.R
import de.xikolo.config.GlideApp
import de.xikolo.controllers.base.CollapsingToolbarViewModelActivity
import de.xikolo.controllers.dialogs.CreateTicketDialog
import de.xikolo.controllers.dialogs.CreateTicketDialogAutoBundle
import de.xikolo.extensions.observe
import de.xikolo.models.Channel
import de.xikolo.utils.extensions.shareCourseLink
import de.xikolo.viewmodels.channel.ChannelViewModel

class ChannelDetailsActivity : CollapsingToolbarViewModelActivity<ChannelViewModel>() {

    companion object {
        val TAG: String = ChannelDetailsActivity::class.java.simpleName
    }

    @AutoBundleField
    lateinit var channelId: String

    //-1 do not scroll to Course
    @AutoBundleField(required = false)
    var scrollToCoursePosition = -1

    override fun createViewModel(): ChannelViewModel {
        return ChannelViewModel(channelId)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel.channel
            .observe(this) {
                setupView(it)
            }
    }

    private fun setupView(channel: Channel) {
        title = channel.title

        val color = channel.colorOrDefault
        collapsingToolbar.setContentScrimColor(color)
        collapsingToolbar.setBackgroundColor(color)
        collapsingToolbar.setStatusBarScrimColor(color)

        if (channel.stageStream?.hlsUrl != null ||
            channel.stageStream?.hdUrl != null ||
            channel.stageStream?.sdUrl != null
        ) {
            imageView.visibility = View.GONE
            lockCollapsingToolbar(channel.title)
        } else if (channel.imageUrl != null) {
            imageView.visibility = View.VISIBLE
            GlideApp.with(this).load(channel.imageUrl).into(imageView)
        } else {
            imageView.visibility = View.GONE
            lockCollapsingToolbar(channel.title)
        }

        val tag = "content"
        val fragmentManager = supportFragmentManager
        if (fragmentManager.findFragmentByTag(tag) == null) {
            val fragment = ChannelDetailsFragmentAutoBundle.builder(channelId)
                .scrollToCoursePosition(scrollToCoursePosition).build()
            val transaction = fragmentManager.beginTransaction()
            transaction.replace(R.id.content, fragment, tag)
            transaction.commit()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.helpdesk, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home    -> {
                NavUtils.navigateUpFromSameTask(this)
                return true
            }
            R.id.action_share    -> {
                shareCourseLink(channelId)
                return true
            }
            R.id.action_helpdesk -> {
                val dialog = CreateTicketDialogAutoBundle.builder().build()
                dialog.show(supportFragmentManager, CreateTicketDialog.TAG)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

}
